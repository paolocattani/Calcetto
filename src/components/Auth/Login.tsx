import { useInput } from '../core/hooks/InputHook';
import { Form, Button, Row, Col } from 'react-bootstrap';
import React from 'react';
import { useHistory, RouteComponentProps, withRouter } from 'react-router-dom';
import { SessionAction, TournamentAction } from 'redux/actions';
import { useDispatch } from 'react-redux';
import { UserDTO } from 'redux/models/user.model';
import { UserMessage } from 'redux/models';
import { toast } from 'react-toastify';

interface PropsType extends RouteComponentProps {}

// https://medium.com/@faizanv/authentication-for-your-react-and-express-application-w-json-web-tokens-923515826e0#6563
const Login: React.FC<PropsType> = (): JSX.Element => {
  const dispatch = useDispatch();
  const currentHistory = useHistory();

  const { value: username, bind: bindUsername } = useInput<string>('');
  const { value: password, bind: bindPassword } = useInput<string>('');

  const handleSubmit = async (evt: React.SyntheticEvent) => {
    evt.preventDefault();
    if (!username || username === '') {
      toast.error('Inserire username o password');
      return;
    }
    if (!password || password === '') {
      toast.error('Inserire la password');
      return;
    }
    try {
      const response = await fetch('/api/v1/auth/authenticate', {
        method: 'POST',
        body: JSON.stringify({ username, password }),
        headers: { 'Content-Type': 'application/json' },
      });
      const result: UserDTO = await response.json();
      if (response.ok && result) {
        // Messaggio di benvenuto
        const message: UserMessage = { type: 'success', message: `Bentornato ${result.username} !` };
        toast.success(message.message);
        dispatch(SessionAction.updateSession({ user: result, message, showMessage: true }));
        dispatch(TournamentAction.fetchTournaments.request({}));
        dispatch(
          SessionAction.sessionControl.request({
            history: currentHistory,
          })
        );
        currentHistory.push('/');
      } else {
        if (response.status === 401) toast.error('Utente o Password errata');
        else toast.error('Errore server. Riprovare piu tardi');
      }
    } catch (error) {
      console.log('onSubmitLogin : ', error);
      toast.error('Errore durante il processo di registrazione. Riprovare piu tardi');
    }
  };

  return (
    <Form onSubmit={handleSubmit}>
      <Form.Group controlId="username">
        <Form.Label>Username o Email</Form.Label>
        <Form.Control required type="text" placeholder="username o email" {...bindUsername} />
      </Form.Group>

      <Form.Group controlId="password">
        <Form.Label>Password</Form.Label>
        <Form.Control required type="password" placeholder="Password" {...bindPassword} />
      </Form.Group>
      <Row>
        <Col>
          <Button size="lg" variant="outline-success" className="float-right" type="submit">
            Conferma
          </Button>
        </Col>
      </Row>
    </Form>
  );
};

export default withRouter(Login);
