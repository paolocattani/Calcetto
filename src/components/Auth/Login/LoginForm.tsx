import * as React from 'react';
import { Form, Field, ErrorMessage } from 'formik';
import { Row, Col, FormControl, Button } from 'react-bootstrap';

const LoginForm: React.FC = () => (
  <Form>
    <Row>
      <Col>
        <label htmlFor="email">Email</label>
        <Field as={FormControl} id="email" name="email" placeholder="Email" type="email" />
        <ErrorMessage name="email" />
      </Col>
    </Row>

    <Row>
      <Col>
        <label htmlFor="password">Password</label>
        <Field as={FormControl} id="password" name="password" placeholder="Password" type="password" />
        <ErrorMessage name="password" />
      </Col>
    </Row>
    <Row>
      <Col>
        <Button type="submit">Login</Button>
      </Col>
    </Row>
  </Form>
);

export default LoginForm;
