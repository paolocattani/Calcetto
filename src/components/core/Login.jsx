// React
import React, { Component } from 'react';
import { Form, Button, Alert } from 'react-bootstrap';
// Router
import { withRouter } from 'react-router';

const formStyle = { width: 'fit-content', margin: 'auto' };

class LoginClass extends Component {
   constructor(props) {
      super(props);
      this.state = { username: '', password: '' };
      this.handleChange = this.handleChange.bind(this);
      this.handleSubmit = this.handleSubmit.bind(this);
   }

   componentDidMount() {
      console.log('Login.componentDidMount : props -> ', this.props);
   }

   handleChange(event) {
      this.setState({ [event.currentTarget.name]: event.currentTarget.value });
   }

   // chiamare API autenticazione
   async handleSubmit(event) {
      event.preventDefault();
      // Test Auth
      const { username, password } = this.state;
      console.log('Login.handleSubmit.userId', username);
      console.log('Login.handleSubmit.password', password);

      const response = await fetch('/api/alfresco/auth', {
         method: 'POST',
         headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
         body: JSON.stringify({ username, password })
      });
      const result = await response.json();
      const { handleLogin } = this.props;
      handleLogin(result.userId, result.id);
      // https://reacttraining.com/react-router/web/example/auth-workflow
      const { history } = this.props;
      const { location } = this.props;
      const { from } = location.state || { from: { pathname: '/' } };
      history.replace(from);
      // this.props.history.push("/");
   }

   // Render form username/password
   render() {
      const message = this.props.location?.state?.message ?? 'Effettua il login per continuare';
      return (
         <>
            {message ? (
               <Alert key={1} variant="danger">
                  {message}
               </Alert>
            ) : null}
            <Form style={formStyle} onSubmit={this.handleSubmit}>
               <Form.Group controlId="username">
                  <Form.Label>Username</Form.Label>
                  <Form.Control name="username" type="text" placeholder="username" onChange={this.handleChange} />
               </Form.Group>

               <Form.Group controlId="password">
                  <Form.Label>Password</Form.Label>
                  <Form.Control name="password" type="password" placeholder="Password" onChange={this.handleChange} />
               </Form.Group>
               <Button variant="warning" type="submit">
                  {' '}
                  Login{' '}
               </Button>
            </Form>
         </>
      );
   }
}

const Login = withRouter(LoginClass);
export default Login;
