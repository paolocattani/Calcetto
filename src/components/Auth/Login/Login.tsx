import * as React from 'react';

import { Formik } from 'formik';
import * as yup from 'yup';
import LoginForm from './LoginForm';
import { ILoginFormValue, ILoginForm } from '../types';

const initialValues: ILoginFormValue = {
  email: '',
  password: ''
};

const validationSchema = yup.object().shape({ password: yup.string(), email: yup.string() });

const Login: React.FC<ILoginForm> = ({ onSubmit }: ILoginForm) => (
  <Formik validationSchema={validationSchema} initialValues={initialValues} onSubmit={onSubmit}>
    <LoginForm />
  </Formik>
);

export default Login;
