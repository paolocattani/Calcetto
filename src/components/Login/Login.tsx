import React from 'react';
import { useSessionContext } from '../core/SessionContext';
import { useHistory } from 'react-router';

// TODO:
export const Login: React.FC = () => {
  const [sessionContext, updateSessionContext] = useSessionContext();
  const currentHistory = useHistory();

  const handleClick = () => {
    updateSessionContext({ ...sessionContext, isAuthenticated: true });
    currentHistory.push('/');
  };

  return <button onClick={handleClick}>Login</button>;
};
