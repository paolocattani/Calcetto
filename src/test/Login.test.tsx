import React from 'react';
import { shallow, mount } from 'enzyme';
import Login from '../components/Auth/Login';
import { shallowToJson } from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import { initialState } from './commons';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

const mockStore = configureStore([]);
describe('<Login />', () => {
  let store;
  let component: any;
  beforeEach(() => {
    store = mockStore(initialState);
    component = shallow(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/']}>
          <Login />
        </MemoryRouter>
      </Provider>
    );
  });

  it('Should render with initial store state', () => {
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should dispatch an action on button click', () => {
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});

test('<Login />', () => {
  const component = shallow(<Login />);
  expect(shallowToJson(component)).toMatchSnapshot();
});
