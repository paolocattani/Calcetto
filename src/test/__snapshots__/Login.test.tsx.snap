// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`<Login /> Should render with initial store state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/",
      ]
    }
  >
    <withRouter(Login) />
  </MemoryRouter>
</ContextProvider>
`;

exports[`<Login /> should dispatch an action on button click 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/",
      ]
    }
  >
    <withRouter(Login) />
  </MemoryRouter>
</ContextProvider>
`;
